src/metabase/pulse/markdown.clj

```diff
@@ -0,0 +1,305 @@
+  [{:keys [tag attrs content]}]
+  (let [resolved-content (if (string? content)
+                           (escape-text content)
+                           (map #(if (string? %)
+                                   (escape-text %)
+                                   (ast->mrkdwn %))
+                                content))
+        joined-content   (str/join resolved-content)]
+    (case tag
+      :document
+      joined-content
+
+      :paragraph
+      (str joined-content "\n")
+
+      :soft-line-break
+      " "
+
+      :hard-line-break
+      "\n"
+
+      (:heading :bold)
+      (str "*" joined-content "*")
+
+      :italic
+      (str "_" joined-content "_")
+
+      :code
+      (str "`" joined-content "`")
+
+      :codeblock
+      (str "```\n" joined-content "```")
+
+      :blockquote
+      (let [lines (str/split-lines joined-content)]
+        (str/join "\n" (map #(str ">" %) lines)))
+
+      :link
+      (let [resolved-uri (resolve-uri (:href attrs))]
+        (if (contains? #{:image :image-ref} (:tag (first content)))
+          ;; If this is a linked image, add link target on separate line after image placeholder
+          (str joined-content "\n(" resolved-uri ")")
+          (str "<" resolved-uri "|" joined-content ">")))
+
+      :link-ref
+      (let [resolved-uri (resolve-uri (-> attrs :reference :attrs :url))]
+        (str "<" resolved-uri "|" joined-content ">"))
+
+      :auto-link
+      (str "<" (:href attrs) ">")
+
+      :mail-link
+      (str "<" (:address attrs) ">")
+
+      ;; li tags might have nested lists or other elements, which should have their indentation level increased
+      (:unordered-list-item :ordered-list-item)
+      (let [indented-content (->> (rest resolved-content)
+                                  str/join
+                                  str/split-lines
+                                  (map #(str "    " %))
+                                  (str/join "\n"))]
+        (if-not (str/blank? indented-content)
+          (str (first resolved-content) indented-content "\n")
+          joined-content))
+
+      :unordered-list
+      (str/join (map #(str "• " %) resolved-content))
+
+      :ordered-list
+      (str/join (map-indexed #(str (inc %1) ". " %2) resolved-content))
+
+      ;; Replace images with text that links to source, including alt text if available
+      :image
+      (let [{:keys [src alt]} attrs]
+        (if (str/blank? alt)
+          (str "<" src "|[Image]>")
+          (str "<" src "|[Image: " alt "]>")))
+
+      :image-ref
+      (let [src (-> attrs :reference :attrs :url)
+            alt joined-content]
+        (if (str/blank? alt)
+          (str "<" src "|[Image]>")
+          (str "<" src "|[Image: " alt "]>")))
+
+      :html-entity
+      (some->> content
+               (get html-entities)
+               (:characters))
+
+      joined-content)))
+
+(defmulti process-markdown
```

Comment:
Made this a multimethod to prepare for adding an implementation for `:email` in a subsequent PR